scp-dbus-service.py

```diff
@@ -162,6 +163,26 @@ def _disconnect_signals (self):
         for s in self._signals:
             g_ppds.disconnect (s)
 
+    def _needs_querying(self, ppdname):
+        # No need to query if we already tried downloading a driver
+        if self.download_tried:
+            return False
+
+        # In EOS we force querying OpenPrinting.org even if an 'exact' fit is
+        # available, unless it's a driver that we already downloaded before via
+        # eos-config-printer, or one of those we chose to pre-install in the OS.
+        regexps = [
+            'lsb/.*eos-config-printer.*ppd.*',   # automatically downloaded via e-c-p
+            'lsb/.*OpenPrinting.*ppd.*',         # e.g. openprinting-ppds-postscript-brother
+            'lsb/.*epson-inkjet-printer.*ppd.*', # e.g. epson-inkjet-printer-201401w
+        ]
```

Comment:
Thanks for the feedback @cosimoc, I've just updated this pull request based on your feedback, see my comments below for more detail on what exactly I changed:

> Not a huge fan of this hardcoded list of strings tbh. What makes the previous logic conclude that there is an exact fit? Could you introduce another fit that is stronger than exact and move the heuristic elsewhere? (Sorry if this does not make sense given the architecture of s-c-p, I'm not too familiar with the internals.)

That was my first approach too, as I was expecting I could perhaps extend the grammar from the .xml file to add the notion of "drivers that, if available locally, don't require querying external sources", but then I realized that is not really the same thing, and should not be mixed:

The 'fit' for a PPD file simply determines how good of a choice it is for a given printer, so that it can be used when creating an ordered list of preferred printer drivers based in the heuristics defined in that file. The new "needs querying" predicate I'm introducing is, however, a completely different thing: it just 'hints' us into knowing whether the locally available PPD file is the best we can get without having to spawn a remote query, downloading a printer driver and inspecting it just to see if it would fit.

So, there are different things, and can't be mixed :-). This said, I re-worked the pull request so that it hopefully sucks a bit less, by doing the following changes:
  * Replaced the "needs querying from OpenPrinting" predicate more generic ("needs querying from external sources"). Might be useful for the future if we accept drivers from other sources.
  * Moved the additional logic out of scp-dbus-printer into the cupshelpers (so that s-c-p's UI could use it if we wanted to -we don't care at the moment, though) and made it look into the whole list of candidate PPD files, not just the best option (if the best option happens not to be a previously downloaded driver from remote sources, that should not spawn a query either).
  * Now that the predicate is not just "remote drivers from OpenPrinting.org", I added the LSB Canon drivers that we pack and pre-install to the list as well, for consistence (and tested with my MG4150) 
  * Added some debug logging

What I did not remove yet, though, is the hardcoded list of regular expressions, but at least it's now inside the cupshelpers together with another bunch of hardcoded heuristics, so it's hopefully better.